src/typescript-worker/playground-typescript-worker.ts

```diff
@@ -90,34 +120,50 @@ const workerAPI: TypeScriptWorkerAPI = {
         },
       ],
     };
-    const diagnostics = new Map<string, lsp.Diagnostic[]>();
-    for (const file of files) {
-      if (file.name.endsWith('.ts') || file.name.endsWith('.js')) {
-        const url = new URL(file.name, self.origin).href;
-        const sourceFile = program!.getSourceFile(url);
-        diagnostics.set(
+
+    const emittedFiles = new Map<string, string>();
+    const fastDiagnostics = new Map<string, lsp.Diagnostic[]>();
+    for (const {file, url} of inputFiles) {
+      fastDiagnostics.set(
+        file.name,
+        languageService.getSyntacticDiagnostics(url).map(makeLspDiagnostic)
+      );
+      const sourceFile = program.getSourceFile(url);
+      program!.emit(
+        sourceFile,
+        (fileName: string, data: string) => {
+          emittedFiles.set(fileName, data);
+        },
+        undefined,
+        undefined,
+        transformers
+      );
+    }
+
+    (async () => {
+      const slowDiagnostics = new Map<string, lsp.Diagnostic[]>();
+      for (const [specifier, content] of await typesFetcher.getFiles()) {
+        // TypeScript is going to look for these files as paths relative to our
+        // source files, so we need to add them to our filesystem with those
+        // URLs.
+        const url = new URL(`node_modules/${specifier}`, self.origin).href;
+        loadedFiles.set(url, {
+          status: 'resolved',
+          content,
+        });
+      }
+      for (const {file, url} of inputFiles) {
```

Comment:
Do you need to bail based on the compileID here too?

Reply:
We could support cancellation here as an optimization, but it's not a correctness issue. The callback on the project side will reject a stale result.

I think it's probably not a big deal for now. Most of the time is spent fetching files, which will probably all be needed for a subsequent compile anyway, so it mostly won't be that big a waste of time.